client/service/src/task_manager.rs

```diff
@@ -74,17 +71,36 @@ impl SpawnTaskHandle {
 		if let Some(metrics) = &self.metrics {
 			metrics.tasks_spawned.with_label_values(&[name]).inc();
 			// We do a dummy increase in order for the task to show up in metrics.
-			metrics.tasks_ended.with_label_values(&[name]).inc_by(0);
+			metrics.tasks_ended.with_label_values(&[name, "normal"]).inc_by(0);
 		}
 
 		let future = async move {
 			if let Some(metrics) = metrics {
-				let poll_duration = metrics.poll_duration.with_label_values(&[name]);
-				let poll_start = metrics.poll_start.with_label_values(&[name]);
-				let task = prometheus_future::with_poll_durations(poll_duration, poll_start, task);
+				// Add some wrappers around `task`.
+				let task = {
+					let poll_duration = metrics.poll_duration.with_label_values(&[name]);
+					let poll_start = metrics.poll_start.with_label_values(&[name]);
+					let inner = prometheus_future::with_poll_durations(poll_duration, poll_start, task);
+					// The logic of `AssertUnwindSafe` here is ok considering that we throw
+					// away the `Future` after it has panicked.
+					panic::AssertUnwindSafe(inner).catch_unwind()
+				};
 				futures::pin_mut!(task);
-				let _ = select(on_exit, task).await;
-				metrics.tasks_ended.with_label_values(&[name]).inc();
+
+				match select(on_exit, task).await {
```

Comment:
why don't you use `select!` for that?

Reply:
`select!` has this additional `FusedFuture` requirement since it's expected to be called in a loop.
I could also add `.fuse()` and use `select!` if that's better, but I don't see this as a tremendous improvement.

Reply:
No, just curious